custom/sensors/PDP.java

```diff
@@ -75,38 +126,89 @@ public double getVoltage() {
 	 *         If PDP connection is lost, InvalidSensorException will be thrown.
 	 */
 	public double getVoltageSafely() throws InvalidSensorException {
-		byte[] rawArray = status3.read();
-		if (rawArray != null) {
-			int rawVoltage = rawArray[6] & 0xFF; // busVoltage is the 7th byte of the third status
-			cachedVoltage = (rawVoltage) * 0.05 + 4.0;
-			lastRead = System.currentTimeMillis();
-		}
+		readStatus3();
 		if (System.currentTimeMillis() - lastRead > PDP.MAX_AGE) {
 			throw new InvalidSensorException("Can not read voltage from PDP");
 		}
 		return cachedVoltage;
 	}
 
+	public double getBatteryResistance() {
+		try {
+			return getBatteryResistanceSafely();
+		}
+		catch (InvalidSensorException e) {
+			LogKitten.ex(e);
+			return cachedResistance;
+		}
+	}
+
+	public double getBatteryResistanceSafely() throws InvalidSensorException {
+		readStatus3();
+		if (System.currentTimeMillis() - lastRead > PDP.MAX_AGE) {
+			throw new InvalidSensorException("Can not read battery resistance from PDP");
+		}
+		return cachedResistance;
+	}
+
+	public double getTotalCurrent() {
+		try {
+			return getTotalCurrentSafely();
+		}
+		catch (InvalidSensorException e) {
+			LogKitten.ex(e);
+			return cachedCurrent;
+		}
+	}
+
 	/**
-	 * Gets the total amperage used by the entire robot.
+	 * Gets the total current used by the entire robot.
 	 *
 	 * @return
-	 * 		Total robot amperage used.
+	 * 		Total robot current used.
 	 * @throws InvalidSensorException
 	 *         If PDP connection is lost, InvalidSensorException will be thrown.
 	 */
-	public double getAmperage() throws InvalidSensorException {
+	public double getTotalCurrentSafely() throws InvalidSensorException {
 		byte[] rawArray = statusEnergy.read();
 		if (rawArray != null) {
 			int rawAmperage = rawArray[1] & 0xFF; // first part of current is the 2nd byte of the energy status
 			rawAmperage = rawAmperage << 4;
 			rawAmperage += (rawArray[2] & 0xF0) >> 4; // second part of current is in the 3rd byte of the energy status
-			cachedAmperage = (rawAmperage) * 0.125;
+			cachedCurrent = (rawAmperage) * 0.125;
 			lastRead = System.currentTimeMillis();
 		}
 		if (System.currentTimeMillis() - lastRead > PDP.MAX_AGE) {
 			throw new InvalidSensorException("Can not read amperage from PDP");
 		}
-		return cachedAmperage;
+		return cachedCurrent;
+	}
+
+	/**
+	 * Gets the current used by a single channel.
+	 *
+	 * @param channel
+	 *        the channel to read the current for
+	 * @return
+	 * 		the current from that channel
+	 * @throws InvalidSensorException
+	 *         if the PDP is not sending data
+	 */
+	public double getCurrentSafely(int channel) throws InvalidSensorException {
+		if (channel < 0) {
+			return 0.0;
+		} else if (channel <= 5) {
+			readStatus(1);
+		} else if (channel <= 11) {
+			readStatus(2);
+		} else if (channel <= 15) {
+			readStatus(3);
+		} else {
+			return 0.0;
```

Comment:
I think throwing some invalid input exception would be most appropriate, but that could cause issues. Technically the current on a channel above 15 will be 0.